rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,198 @@
+        configClazz = TbDeviceStateNodeConfiguration.class,
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeviceStateConfig"
+)
+public class TbDeviceStateNode implements TbNode {
+
+    private static final Set<TbMsgType> SUPPORTED_EVENTS = EnumSet.of(
+            TbMsgType.CONNECT_EVENT, TbMsgType.ACTIVITY_EVENT, TbMsgType.DISCONNECT_EVENT, TbMsgType.INACTIVITY_EVENT
+    );
+    private static final Duration ONE_SECOND = Duration.ofSeconds(1L);
+    private static final Duration ENTRY_EXPIRATION_TIME = Duration.ofDays(1L);
+
+    private Stopwatch stopwatch;
+    private ConcurrentMap<DeviceId, Duration> lastActivityEventTimestamps;
+    private TbMsgType event;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        TbMsgType event = TbNodeUtils.convert(configuration, TbDeviceStateNodeConfiguration.class).getEvent();
+        if (event == null) {
+            throw new TbNodeException("Event cannot be null!", true);
+        }
+        if (!SUPPORTED_EVENTS.contains(event)) {
+            throw new TbNodeException("Unsupported event: " + event, true);
+        }
+        this.event = event;
+        lastActivityEventTimestamps = new ConcurrentHashMap<>();
+        stopwatch = Stopwatch.createStarted();
+        scheduleCleanupMsg(ctx);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) {
+        if (msg.isTypeOf(TbMsgType.DEVICE_STATE_STALE_ENTRIES_CLEANUP_MSG)) {
+            removeStaleEntries();
+            scheduleCleanupMsg(ctx);
+            ctx.ack(msg);
+            return;
+        }
+
+        if (!EntityType.DEVICE.equals(msg.getOriginator().getEntityType())) {
+            ctx.tellSuccess(msg);
+            return;
+        }
+
+        DeviceId originator = new DeviceId(msg.getOriginator().getId());
+
+        lastActivityEventTimestamps.compute(originator, (__, lastEventTs) -> {
+            Duration now = stopwatch.elapsed();
+
+            if (lastEventTs == null) {
+                sendEvent(ctx, originator, msg);
+                return now;
+            }
+
+            Duration elapsedSinceLastEventSent = now.minus(lastEventTs);
+            if (elapsedSinceLastEventSent.compareTo(ONE_SECOND) < 0) {
+                ctx.tellSuccess(msg);
+                return lastEventTs;
+            }
+
+            sendEvent(ctx, originator, msg);
+
+            return now;
+        });
+    }
+
+    private void scheduleCleanupMsg(TbContext ctx) {
+        TbMsg cleanupMsg = ctx.newMsg(
+                null, TbMsgType.DEVICE_STATE_STALE_ENTRIES_CLEANUP_MSG, ctx.getSelfId(), TbMsgMetaData.EMPTY, TbMsg.EMPTY_STRING
+        );
+        ctx.tellSelf(cleanupMsg, Duration.ofHours(1L).toMillis());
```

Comment:
move `Duration.ofHours(1L).toMillis()` to constant.

Reply:
Moved in https://github.com/thingsboard/thingsboard/pull/9030/commits/691a6b5d6a1cc0acd300eca5d9396e296207bf82